### PR TITLE
Adjust to starknet 0.11

### PR DESCRIPTION
**Pull request by THenry14**
## Describe your changes

Adjusted sdk to starknet 0.11
Added support for rpc 0.3.0
Added support for cairo1 hashes calculation
Deprecated gateway provider
Removed WA for docs

## Linked issues

Closes https://github.com/software-mansion/starknet-jvm/issues/249
Closes https://github.com/software-mansion/starknet-jvm/issues/246
Closes https://github.com/software-mansion/starknet-jvm/issues/239
Closes https://github.com/software-mansion/starknet-jvm/issues/225
Closes https://github.com/software-mansion/starknet-jvm/issues/223
Closes https://github.com/software-mansion/starknet-jvm/issues/193

## Breaking changes

- [X] This issue contains breaking changes

Changed interfaces and behaviour of some of the classes/functions

### CODE REVIEW

File: lib/src/main/kotlin/com/swmansion/starknet/crypto/Poseidon.kt [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+            if (size % r == 1) add(Felt.ZERO)
+        }
+
+        var state = Array(m) { longArrayOf(0, 0, 0, 0) }
+
+        for (iter in 0 until inputValues.size step 2) {
+            state = hades(
+                arrayOf(
+                    splitBigInteger(unsplitBigInteger(state[0]) + BigInteger(inputValues[iter].decString())),
+                    splitBigInteger(unsplitBigInteger(state[1]) + BigInteger(inputValues[iter + 1].decString())),
```
Comment (bartekryba) [inline lib/src/main/kotlin/com/swmansion/starknet/crypto/Poseidon.kt:138]: I'm wondering if maybe these additions should be done in the prime field?

Of course it's a very edge case, and maybe there are some boundaries set in the algorithm itself.

File: lib/src/main/kotlin/com/swmansion/starknet/data/types/Contracts.kt [hunk 2]
```diff
@@ -109,24 +116,222 @@ data class ContractDefinition(private val contract: String) {
...
+        @SerialName("CONSTRUCTOR")
+        val constructor: List<DeprecatedCairoEntryPoint>,
+
+        @SerialName("EXTERNAL")
+        val external: List<DeprecatedCairoEntryPoint>,
+
+        @SerialName("L1_HANDLER")
+        val l1Handler: List<DeprecatedCairoEntryPoint>,
+    )
+}
```
Comment (bartekryba) [inline lib/src/main/kotlin/com/swmansion/starknet/data/types/Contracts.kt:247]: In the newest starknet-spec (and devnet), `FunctionAbiEntry` also has an optional `stateVisibility` field, with only one allowed value, `view`.

**Review by bartekryba — COMMENTED** [review]
It looks very good. I went through it carefully, and it caught my eye only in a few places.

File: crypto/poseidon/bindings/starknet_crypto_Poseidon.c [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    jclass longArrayClass = (*env)->FindClass(env, "[J");
+    jobjectArray result = (*env)->NewObjectArray(env, 3, longArrayClass, NULL);
+    for (int i = 0; i < 3; i++) {
+        jlongArray row = (*env)->NewLongArray(env, 4);
+        jlong* elements = (*env)->GetLongArrayElements(env, row, NULL);
+        for (int j = 0; j < 4; j++) {
+            elements[j] = (jlong)converted_values[i][j];
+        }
+        (*env)->ReleaseLongArrayElements(env, row, elements, 0);
+        (*env)->SetObjectArrayElement(env, result, i, row);
```
[the remark this answers is not in the document]
Comment (THenry14) [reply crypto/poseidon/bindings/starknet_crypto_Poseidon.c:35]: To the best of my knowledge, it is safe. 

Although full disclaimer - I'm no expert in that, and mostly tried to learn coding jni on the job here.

File: lib/src/main/kotlin/com/swmansion/starknet/crypto/Poseidon.kt (hunk 1 above)
Comment (THenry14) [reply lib/src/main/kotlin/com/swmansion/starknet/crypto/Poseidon.kt:138]: The C implementation from submodule specifies the prime field and makes sure it operates within its boundries :+1:

File: lib/src/main/kotlin/com/swmansion/starknet/data/types/Contracts.kt (hunk 2 above)
Comment (THenry14) [reply lib/src/main/kotlin/com/swmansion/starknet/data/types/Contracts.kt:255]: Oh, good catch, thanks. I have added this in newest commit + bumped devnet version

**Pull request merged**